proof/invariant-abstract/Syscall_AI.thy

```diff
@@ -475,16 +475,16 @@ lemma do_reply_invs[wp]:
   done
 
 lemma pinv_invs[wp]:
-  "\<lbrace>\<lambda>s. invs s \<and> ct_active s \<and> valid_invocation i s \<and>
-        scheduler_action s = resume_cur_thread\<rbrace>
+  "\<lbrace>\<lambda>s. invs s \<and> ct_active s \<and> valid_invocation i s
+        \<and> bound_sc_tcb_at (\<lambda>a. \<exists>y. a = Some y) (cur_thread s) s\<rbrace>
```

Comment:
This new condition arises from a precondition in the `invs` rule for `send_ipc`, which wants to know that if the `can_donate` flag is true, then the `thread` argument has an sc. It seems like the `thread` argument is (currently) always the current thread (and I think we can discharge this precondition since the current thread is always bound to the current sc when `send_ipc` is called, though I'm not totally sure on this). It would make more sense to me for us to perform some sort of check before calling (a function which calls) `send_ipc`, but that's not how things currently are...

Reply:
I assume `pinv_invs` is called further up in the chain. If you managed to repair those proof, then the condition is fine.

I would however suggest to write `(\<lambda>a. a ~= None)` in the statement instead. `wpsimp` and friends will turn this into the "exists" form automatically.